// hexo-generator-feed/lib/generator.js
'use strict';

var nunjucks = require('nunjucks');
var env = new nunjucks.Environment();
var pathFn = require('path');
var fs = require('fs');

env.addFilter('uriencode', function(str) {
  return encodeURI(str);
});

var atomTmplSrc = pathFn.join(__dirname, '../atom.xml');
var atomTmpl = nunjucks.compile(fs.readFileSync(atomTmplSrc, 'utf8'), env);
var rss2TmplSrc = pathFn.join(__dirname, '../rss2.xml');
var rss2Tmpl = nunjucks.compile(fs.readFileSync(rss2TmplSrc, 'utf8'), env);

module.exports = function(locals) {
  var config = this.config;
  var feedConfig = config.feed;
  var template = feedConfig.type === 'rss2' ? rss2Tmpl : atomTmpl;

  var posts = locals.posts.sort('-date');
  if (feedConfig.limit) posts = posts.limit(feedConfig.limit);

  var url = config.url;
  if (url[url.length - 1] !== '/') url += '/';

  var gl = function(content){
    var str = content.replace(/<figure[^>]*>[\s\S]*?<\/figure>/g,function(str){
      return str.match(/<pre>[\s\S]*?<\/pre>/g)[1]
    });
    str = str.replace(/<span class="(strong|emphasis|meta-string|meta|addition|code|bullet|built_in|literal|keyword|attribute|selector-tag|meta-keyword|doctag|name|type|string|number|selector-id|selector-class|quote|template-tag|deletion|title|section|regexp|selector-pseudo|selector-attr|link|template-variable|variable|symbol|tag|function|comment|attr|undefined|subst|params)">([\s\S]*?)<\/span>/g,"$2");
    str = str.replace(/<span class="(keyword|class|stylus|name|sqf|javascript|xml|ruby)">([\s\S]*?)<\/span>/g,"$2")
  	return str;
  }
  /* 防爬虫 */
  var f = function(url){
    return '<p>欢迎来我的博客<a href="' + url + '">查看原文</a>。</p>'
  }

  var xml = template.render({
    config: config,
    url: url,
    posts: posts,
    feed_url: config.root + feedConfig.path,
    gl: gl,
    f: f
  });

  return {
    path: feedConfig.path,
    data: xml
  };
};
