// hexo-generator-search/lib/generator.js
var ejs = require('ejs');
var pathFn = require('path');
var fs = require('fs');
var moment = require('moment');

ejs.filters.cdata = function(str){
  return str ? '<![CDATA[' + str + ']]>' : '';
};

var searchTmplSrc = pathFn.join(__dirname, '../search.ejs');
var searchTmpl = ejs.compile(fs.readFileSync(searchTmplSrc, 'utf8'));

module.exports = function(locals){
  var config = this.config;
  var searchConfig = config.search;
  var template = searchTmpl;
  var searchfield = searchConfig.field;

  var posts, pages;

  if(searchfield.trim() != ''){
    searchfield = searchfield.trim();
    if(searchfield == 'post'){
      posts = locals.posts.sort('-date');
    }else if(searchfield == 'page'){
      pages = locals.pages;
    }else{
      posts = locals.posts.sort('-date');
      pages = locals.pages;
    }
  }else{
    posts = locals.posts.sort('-date');
  }

  var getTags = function(tags,categories){
    var str = '';
    tags.data.forEach(function(item){
      str +=item.name+' ';
    })
    categories.data.forEach(function(item){
      str +=item.name+' ';
    })
    return str
  }
  var getDate = function(d){
    return moment(d).format('YYYY-MM-DD');
  }
  var xml = template({
    getDate: getDate,
    config: config,
    posts: posts,
    pages: pages,
    getTags: getTags,
    feed_url: config.root + searchConfig.path
  });

  return {
    path: searchConfig.path,
    data: xml
  };
};
